src/DI/ServiceLookup/CallSiteFactory.cs

```diff
@@ -193,26 +193,127 @@ private IServiceCallSite TryCreateExact(ServiceDescriptor descriptor, Type servi
+                    != GenericParameterAttributes.None)
+                {
+                    if (!parameterTypeInfo.IsValueType ||
+                        (parameterTypeInfo.IsGenericType && IsGenericTypeDefinedBy(parameter, typeof(Nullable<>))))
+                    {
+                        return false;
+                    }
+                }
             }
 
-            return null;
+            return true;
+        }
+
+        private static bool IsGenericTypeDefinedBy(Type type, Type openGeneric)
+        {
+            return !type.GetTypeInfo().ContainsGenericParameters
+                       && type.GetTypeInfo().IsGenericType
+                       && type.GetGenericTypeDefinition() == openGeneric;
+        }
+
+        private static Type SubstituteGenericParameterConstraint(Type[] parameters, Type constraint)
+        {
+            if (!constraint.IsGenericParameter) return constraint;
+            return parameters[constraint.GenericParameterPosition];
+        }
+
+        private static bool ParameterCompatibleWithTypeConstraint(Type parameter, Type constraint)
+        {
+            return constraint.IsAssignableFrom(parameter) ||
+                   new[] { parameter, parameter.BaseType }
+                       .Concat(parameter.GetTypeInfo().ImplementedInterfaces)
+                       .Where(p => p != null)
+                       .Any(p => ParameterEqualsConstraint(p, constraint));
+        }
+
+        //[SuppressMessage("Microsoft.Design", "CA1031", Justification = "Implementing a real TryMakeGenericType is not worth the effort.")]
+        private static bool ParameterEqualsConstraint(Type parameter, Type constraint)
+        {
+            var genericArguments = parameter.GenericTypeArguments;
+            if (genericArguments.Length > 0 && constraint.GetTypeInfo().IsGenericType)
+            {
+                var typeDefinition = constraint.GetGenericTypeDefinition();
+                if (typeDefinition.GetTypeInfo().GenericTypeParameters.Length == genericArguments.Length)
+                {
+                    try
+                    {
+                        var genericType = typeDefinition.MakeGenericType(genericArguments);
+                        var constraintArguments = constraint.GetTypeInfo().GenericTypeArguments;
+
+                        for (var i = 0; i < constraintArguments.Length; i++)
+                        {
+                            var constraintArgument = constraintArguments[i].GetTypeInfo();
+                            if (!constraintArgument.IsGenericParameter && !constraintArgument.IsAssignableFrom(genericArguments[i].GetTypeInfo()))
+                            {
+                                return false;
+                            }
+                        }
+
+                        return genericType == parameter;
+                    }
+                    catch (Exception)
```

Comment:
nitpick - this is supposed to catch line#295 right? Specifically for catch-alls, it would be clearer to make the try{} body as specific as possible, so perhaps catch in 296?

Reply:
Good call!